master/buildbot/util/gceclientservice.py

```diff
@@ -0,0 +1,196 @@
+        HTTPClientService.__init__(self, "https://www.googleapis.com")
+        self.sa_credentials = sa_credentials
+        self.scopes = scopes
+        self.renderer = renderer
+        self.project = project
+        self.zone = zone
+        self.instance = instance
+
+        self.token = None
+        self.tokenExpiry = 0
+
+    @defer.inlineCallbacks
+    def getBearerToken(self):
+        if time.time() < self.tokenExpiry:
+            return self.token
+
+        now = time.time()
+        exp = now + 3600
+
+        if self.renderer is not None:
+            sa_credentials = yield self.renderer.renderSecrets(self.sa_credentials)
+        else:
+            sa_credentials = self.sa_credentials
+
+        if isinstance(sa_credentials, str):
+            sa_credentials = json.loads(sa_credentials)
+
+        data = {
+            "iss": sa_credentials['client_email'],
+            "scope": " ".join(self.scopes),
+            "aud": "https://www.googleapis.com/oauth2/v4/token",
+            "iat": now, 'exp': exp
+        }
+        jwt_token = jwt.encode(data, key=sa_credentials['private_key'], algorithm="RS256")
+
+        response = yield self.post("/oauth2/v4/token",
+            needToken=False, data=urllib.parse.urlencode({
+                "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
+                "assertion": jwt_token
+            }),
+            headers={'Content-Type': "application/x-www-form-urlencoded"}
+        )
+        if response.code >= 400:
+            raise Exception("Failed to get authorization token (HTTP code {0}".
+                format(response.code))
+
+        token_result = yield response.json()
+
+        if token_result['token_type'] == 'Bearer':
+            self.tokenExpiry = exp - 60
+            self.token = token_result['access_token']
+            return defer.returnValue(self.token)
+        else:
+            raise "failed to get an access token"
+
+    @defer.inlineCallbacks
+    def _prepareRequest(self, ep, kwargs):
+        url, req_kwargs = HTTPClientService._prepareRequest(self, ep, kwargs)
+
+        needToken = kwargs.pop('needToken', True)
+        if needToken:
+            token = yield self.getBearerToken()
+            req_kwargs['headers']['Authorization'] = "Bearer {0}".format(token)
+
+        return url, req_kwargs
+
+    def zoneEndpoint(self, method=None):
+        base = "/compute/v1/projects/{0}/zones/{1}".format(
+            self.project, self.zone)
+        if method:
+            return "{0}/{1}".format(base, method)
+        else:
+            return base
+
+    def instanceEndpoint(self, method=None):
+        rel = "instances/{0}".format(self.instance)
+        if method:
+            rel = "{0}/{1}".format(rel, method)
+        return self.zoneEndpoint(rel)
+
+    def processRequest(self, deferred):
+        return self.validateRes(deferred)
+
+    @defer.inlineCallbacks
+    def processAsyncRequest(self, deferred):
```

Comment:
I think this name does not reflect the fact that it waits for operationEnd.

I think it would be better API to rather have option waitForFinish option in high level APIs.

Reply:
The point of keeping `processRequest` / `processAsyncRequest` separate is to interleave calls when possible, to reduce the setup time. I found it a lot easier to debug by having it out of each method, as the patterns are clear:

- `<do something>` returns the deferred for the API call
- `processRequest` waits for a sync operation to finish
- `processAsyncRequest` waits for an async operation to finish 

Maybe just renaming them into `waitSyncRequest` and `waitAsyncRequest` ?

Adding `waitForFinish` would lead to a lot of duplicate code, and murky the flow (since in half of the cases, `processAsyncRequest` is done after the API call)

Reply:
I think the sync and async naming is confusing.
In twisted, everything is async
in GCP, some requests return immediately, and create an operation log
some requests do the action within the http request, and finish "immediately".
Your API tries to hide this detail, and I think it shouldn't.

So I would prefer that:
- every API has its status code parsed automatically, and translated into exception (processRequest would be implicit)
- have a separate waitForOperations() API, which you can put when is needed (and which could wait for several operations to finish)

waitForFinish could be implemented as a decorator if you want to avoid my proposed boilerplate.

Reply:
> In twisted, everything is async

Got it ... Wasn't seeing "async" from this angle but from the GCE angle. I'm modifying the code to call validateRes mandatorily in every request and then have waitForOperation to wait for an operation as you suggested.

I'm not going to go the waitForFinish route. Don't want to write the boilerplate and I have spent already a lot of time implementing this, don't want to have to learn about decorators now. Python is not my primary language ;-)